### ISSUE TITLE
12.0.0 issues with 'new-window' event (and maybe also 'ready-to-show'?

### ISSUE THREAD

**Issue by ericblade** [electron/electron#27977]
I know that there were some changes involved with new window, specifically implementing a new event that is triggered before the window creation, rather than intercepting 'new-window' which is now deprecated.  However, 'new-window' itself appears to have some changes.

In my application, I have a preload script that checks the URL that is loaded, and then creates some instances of some different classes, which then associate themselves to the new window.  One of those classes calls 'document.body.appendChild()' inside it's constructor.  Since upgrading to 12.0.0, that class fails, indicating that 'document.body' is undefined.  Reverting to 11.x, this is not an issue.

While I do plan to upgrade to the new event ( https://github.com/electron/electron/pull/24517 ), where I have no idea if this same code will work yet (but I suspect it won't...), this does indicate that something unexpected has changed in the now deprecated 'new-window' event.

Several other hooks inside code for this project were tentatively driven off of setTimeout() due to lack of being able to find an adequate event to drive them, and I've had to significantly increase their timeout lengths (from 1sec to 2 or 3sec) to make them trigger correctly with the pieces required from the remote app in place.  The windows appear to show significantly faster than before, and they are shown on an event inside the node app, driven on the 'ready-to-show' event.

This leads me to believe that something has changed in the timing of these items -- preload code is running before a document.body exists, when previously that was not the case ... perhaps ready-to-show is also firing earlier than before, although it may just be that electron 12 is faster itself, and getting to that point faster, as i haven't noticed any specific code issues related to ready-to-show.. but there's nothing dependent upon that event, either, other than "show"

### Preflight Checklist

* [ x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [ x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [ x] I have searched the issue tracker for an issue that matches the one I want to file, without success.

### Issue Details

* **Electron Version:** 12.0.0
  * 
* **Operating System:** Microsoft Windows [Version 10.0.21318.1000]
  * 
* **Last Known Working Electron version:** 11.x
  *

**Comment by VerteDinde** [comment]
Hey @ericblade, thanks for reporting this issue. Would it be possible for you to make a standalone testcase with only the code necessary to reproduce the issue? For example, Electron Fiddle is a great tool for making small test cases and makes it easy to publish your test case to a gist that Electron maintainers can use.

Stand-alone test cases make fixing issues go more smoothly: it ensure everyone's looking at the same issue, it removes all unnecessary variables from the equation, and it can also provide the basis for automated regression tests.

I'm adding the blocked/needs-repro label for this reason. After you make a test case, please link to it in a follow up comment.

Thanks so much!

**Comment by ericblade** [comment]
hmm.  If I find some time, sure.  Looks like although initial looks at my app indicated that only minor changes would have most of it functional in 12, it looks like it broke a whole lot more, once I started digging into it, so reverting back to 11.

**Comment by ckerr** [comment]
> hmm. If I find some time, sure. ... it looks like it broke a whole lot more, once I started digging into it, so reverting back to 11.

Thank you for reporting this bug, but saying things like ``it broke a whole lot more'' without reporting something actionable is not super useful.

**Comment by ericblade** [comment]
> > hmm. If I find some time, sure. ... it looks like it broke a whole lot more, once I started digging into it, so reverting back to 11.
> 
> Hey, thank you for reporting this issue, but please try to give actionable information. Saying things like ``it broke a whole lot more'' does not give triagers anything actionable or even anything to research.

I understand that, but I also don't have time to figure out what broke or why, if it's simply this event causing a cascading failure, or if there are other problems.  I'm pretty sure that I have provided more than enough info for replicating the problem reported in this issue.

**Comment by TheCleric** [comment]
> In my application, I have a preload script that checks the URL that is loaded, and then creates some instances of some different classes, which then associate themselves to the new window.  One of those classes calls 'document.body.appendChild()' inside it's constructor.  Since upgrading to 12.0.0, that class fails, indicating that 'document.body' is undefined.  Reverting to 11.x, this is not an issue.

I also ran into this issue (https://github.com/nativefier/nativefier/issues/1197). I found it was because, by default, calls to render level `window.open` no longer return a full `BrowserWindow` object but rather a `BrowserWindowProxy` that doesn't have a `document.body`. In my instance I had to use 11.x and set `nativeWindowOpen` in the `webPreferences` of your `BrowserWindow` constructor to true. You might be able to get away with just the `nativeWindowOpen=true` part to fix your issue.

**Comment by nornagon** [comment]
I'm closing this issue because it is not actionable. There is no repro, and the description is insufficiently precise to create one. If a standalone testcase can be produced, I'll be happy to consider reopening.

NB, definitely not saying there's nothing wrong here! You're right that these events changed quite a bit recently, and I believe you that we might have broken something in a subtle way. But there are over 1,000 open issues on the tracker, and about 3 maintainers who regularly triage issues. There is no possible way we can address every issue if we have to reverse-engineer what the bug is every time. So, we ask that you bring a little effort to the table to clearly illustrate the issue you're experiencing, by way of a standalone testcase.

**Comment by ericblade** [comment]
TL;DR summary:

In windows opened via the new-window event, in electron 12.0.0, the preload script document.body is undefined in code running the first tick.

In 11, the preload script was able to access document.body immediately.

This event shouldn't have changed at all, other than to be marked deprecated, in favor of the new event.  If due to the architecture changes to support the new event, the original event cannot be properly recreated to act exactly as it did, then it would be better to just drop it rather than deprecate it.

Honestly, so long as I can continue to use electron 11, and I'm not getting paid to upgrade this project to 12, I couldn't care less if the problem is fixed, but I figured I would offer that there is definitely a major error in how this event is handled in comparison to the previous version.  My client has no interest in spending their money to help solve issues in components they don't own.

**Comment by nornagon** [comment]
> windows opened via the new-window event

This is the bit that's unclear to me. Are you calling `preventDefault` and providing a different webcontents with `newGuest`?

**Comment by ericblade** [comment]
> This is the bit that's unclear to me. Are you calling `preventDefault` and providing a different webcontents with `newGuest`?

As far as I'm aware, that's the only valid way to use the new-window event to provide a preload?  Apologies if I was unintentionally unclear because I don't presently know of any other way to make use of that event.

```
        globalBrowser.webContents.on('new-window', (event, url, frameName, disposition, options, additionalFeatures, referrer, postBody) => {
            event.preventDefault();
            const win = new BrowserWindow({
                webContents: options.webContents,
                webPreferences: {
                    nativeWindowOpen: true,
                    preload: path.join(app.getAppPath(), 'preload.js'),
                    enableRemoteModule: true,
                },
                autoHideMenuBar: true,
                icon: path.join(__dirname, '/../icon.png'),
                show: false,
            });
```

**Comment by nornagon** [comment]
Thanks for the clarification. There are a couple of ways to run preload scripts, including https://www.electronjs.org/docs/api/session#sessetpreloadspreloads.

**Comment by ericblade** [comment]
Does that affect other windows?

![image](https://user-images.githubusercontent.com/1451847/123182262-08d29e80-d45d-11eb-92e4-f143baff6245.png)

The documentation doesn't actually suggest anything about what a Session is.
